File: baekjoon/20190219/11724.andi.py
def main():
  n, m = [int(x) for x in input().strip().split(" ")]
  nodes = [[] for _ in range(n)]
  visited = [False for _ in range(n)]
  for _ in range(m):
    u, v = [int(x) for x in input().strip().split(" ")]
    nodes[u-1].append(v-1)
    nodes[v-1].append(u-1)
  cc_count = 0
  for i, _ in enumerate(nodes):
    if not visited[i]:
      cc_count += 1
    else:
      continue
    dfs(nodes, i, visited)
  print(cc_count)

def dfs(nodes, i, visited):
  if visited[i]:
    return
  visited[i] = True
  for a in nodes[i]:
    dfs(nodes, a, visited)

main()

File: baekjoon/20190219/11724.andi.java
import java.util.*;

class Main {
  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    final String[] split1 = scanner.nextLine().split(" ");
    final int n = Integer.parseInt(split1[0]);
    final int m = Integer.parseInt(split1[1]);
    ArrayList<LinkedList<Integer>> nodes = new ArrayList<>(n);
    boolean[] visited = new boolean[n];
    for (int i = 0; i < n; i++) {
      nodes.add(new LinkedList<>());
      visited[i] = false;
    }

    for (int i = 0; i < m; i++) {
      final String[] split2 = scanner.nextLine().split(" ");
      final int u = Integer.parseInt(split2[0]) - 1;
      final int v = Integer.parseInt(split2[1]) - 1;
      nodes.get(u).add(v);
      nodes.get(v).add(u);
    }

    scanner.close();

    int ccCount = 0;
    for (int i = 0; i < n; i++) {
      if (visited[i]) {
        continue;
      }
      ccCount++;
      dfs(nodes, i, visited);
    }
    System.out.println(ccCount);
  }

  private static void dfs(ArrayList<LinkedList<Integer>> nodes, int i, boolean[] visited) {
    if (visited[i]) {
      return;
    }
    visited[i] = true;
    for (int adj : nodes.get(i)) {
      dfs(nodes, adj, visited);
    }
  }

}